src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -117,26 +102,39 @@ public bool HasReachedMaxErrors
         /// <inheritdoc />
         public int Count
         {
-            get { return _innerDictionary.Count; }
+            get
+            {
+                var count = 0;
+                foreach (var item in this)
+                {
+                    count++;
+                }
+
+                return count;
```

Comment:
It's a bit messy to deal with `Merge`, `MarkFieldValid` (for a previously not-visible node), `Remove` etc. I could do lazily evaluate so any of these operations would clear the count and we can rebuild the count once we request for it.